src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -819,7 +819,7 @@ public SCMRevision run(GitClient client, String remoteName) throws IOException,
                                       if (branches.isEmpty()) {
```

Comment:
The actual bug fix is the line above (need to change `false` to `true`)
```
List<Branch> branches = client.getBranchesContaining(hash, true);
```

paired with some deconstruction of the remote branch name back into the actual corresponding branch name two lines after your change...

```
// something like this... but will need debugging to confirm
String name = branches.stream()
    .map(getName)
    .filter(n->n.startsWith(Constants.R_REFS+context.remoteName()+"/"))
    .map(n->n.substring(Constants.R_REFS.length()+context.remoteName().length()+1))
    .first();
```

The change you have proposed would address the case where the revision is in the repository but not visible from `refs/heads/` (could be visible from some other custom ref or could just not have been GC'd yet) but in that case because the refspec will be constructed as `refs/heads/${revision}` which does not exist, the fetch will fail on checkout of a shared library unless the workspace already has fetched the revision and the `SCMFileSystem` will fail unless the local repository cache already contains the revision.